modules/core/src/lib/widget-manager.ts

```diff
@@ -0,0 +1,237 @@
+  /** Called when a dragend event occurs */
+  onDragEnd?: (info: PickingInfo, event: MjolnirGestureEvent) => void;
+}
+
+export abstract class Widget<PropsT = {}> implements IWidget {
+  // / Populated by core when mounted
+  /** The Deck instance that the widget is attached to */
+  deck?: Deck;
+  /** The HTML element that the widget creates */
+  element?: HTMLDivElement | null;
+  /** The id of the view that the widget is attached to */
+  viewId?: string | null;
+
+  props: PropsT;
+
+  constructor(props: PropsT) {
+    this.props = props;
+  }
+
+  setProps(props: Partial<PropsT>) {
+    this.props = {...this.props, ...props};
+  }
+
+  abstract onAdd(): HTMLDivElement | null;
+
+  onRemove() {}
+}
+
+const PLACEMENTS = {
+  'top-left': {top: 0, left: 0},
+  'top-right': {top: 0, right: 0},
+  'bottom-left': {bottom: 0, left: 0},
+  'bottom-right': {bottom: 0, right: 0},
+  fill: {top: 0, left: 0, bottom: 0, right: 0}
+} as const;
+
+export type WidgetPlacement = keyof typeof PLACEMENTS;
+
+const ROOT_CONTAINER_ID = '__root';
+
+export class WidgetManager {
+  deck: Deck;
+  parent?: HTMLElement | null;
+  containers: {[id: string]: HTMLDivElement} = {};
+  widgets: IWidget[] = [];
+  lastViewports: {[id: string]: Viewport} = {};
+
+  constructor({deck, parent}: {deck: Deck; parent?: HTMLElement | null}) {
+    this.deck = deck;
+    this.parent = parent;
+  }
+
+  finalize() {
+    for (const widget of this.widgets) {
+      this.remove(widget);
+    }
+    for (const id in this.containers) {
+      this.containers[id].remove();
+    }
+  }
+
+  add(
+    widget: IWidget,
+    opts: {
+      viewId?: string | null;
+      placement?: WidgetPlacement;
+    } = {}
+  ) {
+    if (this.widgets.includes(widget)) {
+      // widget already added
+      return;
+    }
+
+    const {placement = 'top-left', viewId = null} = opts;
+    widget.deck = this.deck;
+    widget.viewId = viewId;
+    widget.element = widget.onAdd();
+
+    if (widget.element) {
+      this._getContainer(viewId, placement).append(widget.element);
+    }
+    this.widgets.push(widget);
+  }
+
+  remove(widget: IWidget) {
+    const i = this.widgets.indexOf(widget);
+    if (i < 0) {
+      // widget not found
+      return;
+    }
+    this.widgets.splice(i, 1);
+    widget.onRemove();
+
+    widget.deck = undefined;
+    if (widget.element) {
+      widget.element.remove();
+    }
+    widget.element = undefined;
+  }
+
+  /* global document */
+  private _getContainer(viewId: string | null, placement: WidgetPlacement): HTMLDivElement {
+    const containerId = viewId || ROOT_CONTAINER_ID;
+    let viewContainer = this.containers[containerId];
+    if (!viewContainer) {
+      viewContainer = document.createElement('div');
+      viewContainer.style.pointerEvents = 'none';
+      viewContainer.style.position = 'absolute';
+      viewContainer.style.overflow = 'hidden';
+      this.parent?.append(viewContainer);
```

Comment:
Should updateContainers get called here or in `add` so they have a set size asap? I could imagine a widget expecting its container size to be known during initialization, or soon after.

Reply:
There is no guarantee that the viewport known to `onAdd` is accurate. I would argue that if a widget depends on its parent container size, it needs to subscribe to `onViewportChange`.